data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,441 @@
+WHERE person_id != {pid}
+AND {date_column} < (SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid})
+"""
+CLEAN_QUERY_END_DATE = """
+SELECT *
+FROM `{project}.{dataset}.{table}`
+WHERE person_id != {pid}
+AND (CASE WHEN {end_date_column} IS NOT NULL THEN {end_date_column} < (SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid} ) ELSE CASE WHEN {end_date_column} IS NULL THEN {start_date_column} < (
+SELECT deactivated_date
+FROM `{deactivated_pids_project}.{deactivated_pids_dataset}.{deactivated_pids_table}`
+WHERE person_id = {pid}) END END)
+"""
+# Deactivated participant table fields to query off of
+PID_TABLE_FIELDS = [{
+    "type": "integer",
+    "name": "person_id",
+    "mode": "required",
+    "description": "The person_id to retract data for"
+}, {
+    "type": "date",
+    "name": "deactivated_date",
+    "mode": "required",
+    "description": "The deactivation date to base retractions on"
+}]
+
+
+def get_pids_datasets_and_tables(project_id):
+    """
+    Get list of tables that have a person_id column; along with a separate
+    list of the datasets containing pids
+    """
+    dataset_obj = bq.list_datasets(project_id)
+    datasets = [d.dataset_id for d in dataset_obj]
+
+    pids_table_list = []
+    pids_dataset_list = []
+    for dataset in datasets:
+        pid_tables = sandbox_and_remove_pids.get_tables_with_person_id(
+            project_id, dataset)
+        if pid_tables:
+            pids_table_list.extend(pid_tables)
+            pids_dataset_list.append(dataset)
+
+    return pids_table_list, pids_dataset_list
+
+
+def get_date_info_for_pids_tables(project_id):
+    """
+    Loop through tables within all datasets and determine if the table has an end_date date or a date field. Filtering
+    out person and death table and keeping only tables with PID and an upload or start/end date associated.
+
+    :param project_id: bq name of project_id
+    :return: filtered dataframe which includes the following columns for each table in each dataset with a person_id
+    'project_id', 'dataset_id', 'table', 'date_column', 'start_date_column', 'end_date_column'
+    """
+    logging.info('getting all tables with person id in %s' % project_id)
+    pids_tables, pids_datasets = get_pids_datasets_and_tables(project_id)
+    # Create empty df to append to for final output
+    date_fields_info_df = pd.DataFrame()
+
+    # Loop through datasets
+    logging.info(
+        'Looping through datasets to filter and create dataframe with correct date field to determine retraction'
+    )
+    for dataset in pids_datasets:
+        # Get table info
+        table_info_df = bq.get_table_info_for_dataset(project_id, dataset)
+        # Filter out to only records with pids
+        pids_tables_df = table_info_df[table_info_df['table_name'].isin(
+            pids_tables)]
+
```

Comment:
I think this can be refactored some.  Right now, `get_pids_datasets_and_tables` uses this query:
```
SELECT table_name
FROM `{project}.{dataset}.INFORMATION_SCHEMA.COLUMNS`
WHERE COLUMN_NAME = 'person_id'
```
Then `get_table_info_for_dataset` is using this query:
```
SELECT table_name
FROM `{project}.{dataset}.INFORMATION_SCHEMA.COLUMNS`
```
These are nearly identical BigQuery calls.  You can probably use the results from this second query to populate your lists.  But I would argue that the `pids_tables` list is useless because you are only using it to get the list of tables with a `person_id` and then subselect the tables with a `'DATE'` field.

On the test project, this results in  879 bigquery calls from `get_pids_datasets_and_tables` and 146 bigquery calls in `get_table_info_for_dataset`.  This took 0:29:06.475180 minutes to complete.  If this is something we wish to retain and possibly use again, we should try to make this as clean as possible with as few calls to bigquery as possible.

And this time isn't taking into account any actual retraction queries.  It is only the time for running the queries in these two functions for the test project.

Reply:
I was able to simplify this and only call bigquery to get table information with only tables that have a `person_id` column.